workflow/scripts/spia.R

```diff
@@ -87,6 +93,60 @@ if (nrow(sig_genes) == 0) {
       rename(
         `pathway id` = value
       )
+
+    extract_genes_from_pathway <- function(pathway) {
+      # Extract genes from protEdges and mixedEdges
+      genes_protEdges <- unique(c(pathway@protEdges$src, pathway@protEdges$dest))
+      genes_mixedEdges <- unique(c(pathway@mixedEdges$src[pathway@mixedEdges$src_type == "ENSEMBL"], 
+                                  pathway@mixedEdges$dest[pathway@mixedEdges$dest_type == "ENSEMBL"]))
+      
+      # Combine all gene IDs
+      all_genes <- unique(c(genes_protEdges, genes_mixedEdges))
+      
+      # Map ENSMUSB values to gene names
+      external_gene_names <- mapIds(orgDb, 
+                                    keys = all_genes, 
+                                    column = "SYMBOL", 
+                                    keytype = "ENSEMBL",
+                                    multiVals = "first")
+      
+      # Check if external_gene_names is empty or NULL
+      if (is.null(external_gene_names) || length(external_gene_names) == 0) {
+          return("")
+      }
+
+      # Filter diffexp for relevant gene names
+      filtered_diffexp <- diffexp %>%
+          filter(ext_gene %in% unname(external_gene_names)) %>%
+          select(ext_gene, beta_val = !!beta_col) %>%
+          drop_na()
+      
+      # Check if filtered_diffexp is empty
+      if (nrow(filtered_diffexp) == 0) {
+          return("")
+      }
+
+      # Create result string
+      result_string <- filtered_diffexp %>%
+          mutate(pair = paste(ext_gene, beta_val, sep = ":")) %>%
+          pull(pair) %>%
+          paste(collapse = ", ")
+      
+      return(result_string)
+    }
+
+    extract_genes_from_pathway_v <- Vectorize(extract_genes_from_pathway)
```

Comment:
If this is really necessary, I would go for the more readable:
```suggestion
    extract_genes_from_pathway_v <- str_split(extract_genes_from_pathway, ", ")
```